src/ng/compile.js

```diff
@@ -2424,6 +2426,9 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           if (isFunction(controller.instance.$onInit)) {
             controller.instance.$onInit();
           }
+          if (isFunction(controller.instance.$onDestroy)) {
+            scope.$on('$destroy', controller.instance.$onDestroy.bind(controller.instance));
```

Comment:
The problem with this is that `scope` might be a parent scope that is not directly related to the element.
Maybe we could use `$element.on('$destroy', ...)` instead (I haven't given it much thought, though).